software/controller/lib/core/nvparams.h

```diff
@@ -0,0 +1,89 @@
+/* Copyright 2020, RespiraWorks
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#ifndef NVPARAM_H_
+#define NVPARAM_H_
+
+#include "network_protocol.pb.h"
+#include "units.h"
+#include <stddef.h>
+#include <stdint.h>
+
+// This structure defines the layout of the non-volatile
+// parameter info stored in the I²C EEPROM.
+struct NVparams_struct {
+  // Header info used to keep track of parameter info
+  uint32_t crc{0};    // 32-bit CRC of remaining structure
+  uint8_t count{0};   // Incremented on each write.
+  uint8_t version{0}; // Version of the structure.
+
+  uint8_t reinit{0}; // Write to 1 (through dbg_reinit) to request for a reinit
+                     // on next boot. This should prove useful if our system
+                     // memory ends up in a weird state during testing.
+
+  uint8_t reserved{0}; // kept here for alignment's sake
+
+  uint32_t vent_serial_number{0};
+  // Non-volatile parameters should be added here
+  uint32_t power_cycles{0};      // Count of device ON/OFF cycles.
+  uint32_t cumulated_service{0}; // Cumulated power-ON time, stored in seconds.
+                                 // May rollover after 136 years
+  VentParams last_settings =
+      VentParams_init_default; // Last settings seen by the vent
+};
+
+// We are reserving the first 8 kB out of our 32kB eeprom for nv params.
+// Since we use a flip/flop that means NVparams_struct should be at most 4kB.
+static_assert(sizeof(NVparams_struct) <= 4096);
+
+enum class NVParamsAddress {
+  kFlip = 0,
+  kFlop = 4096,
+};
+
+// Class that encapsulates NVparams. We need the NVparams_struct to be
+// defined independently in order to facilitate access and casting, but
+// the actual struct we use is encapsulated here in order to write the
+// changes made to the struct in EEPROM.
+class NVParams {
+public:
+  NVParams() = default;
+  void Init();
+  bool Set(uint16_t offset, void *value, uint8_t len);
```

Comment:
why reinterpret_cast rather than static_cast out of interest?

Reply:
Because reinterpret_cast takes care of endianness for me. Depending on the processor's endianness, reinterpret_cast<uint8_t*>(&some_int32) will return different pointers.
It shouldn't make any difference when we read and write using the same order, but it may or may not make a difference when writing data that is not aligned to 32bits.

Reply:
so I may have misunderstood, but we are going to convert the void* to a uint8 right? so isn't it better to do that explicitly, to make it clear to the caller that the only valid data is a uint8? if that still causes issues with endianness maybe we could create a  type to represent an endian independent byte?

Reply:
The problem comes from the fact that NVParams.Set can put different data types into the nvparams_struct, we will always need a pointer (so, either void * or uint8_t *) + length of the data. The reason I use void* is to make is easy on the caller who would otherwise pass us reinterpret_cast<uint8_t*>(&data) whenever data isn't a uint8_t (which is basically always).